src/main/java/com/conveyal/osmlib/NodeTracker.java

```diff
@@ -1,84 +1,78 @@
+ * space (i.e. they only need, so far, one more than an int provides), only a few blocks are used.
  */
 public class NodeTracker {
-
     private static final Logger LOG = LoggerFactory.getLogger(NodeTracker.class);
-    
-    /** Lowest 6 bits, index into a single 64-bit long int. */
-    private static final long LOW_MASK = 0x3F; 
-    private static final int LOW_SHIFT = 6;
-    
-    /** Lowest 12 bits, bit position within block. */
-    private static final long MID_MASK = 0xFFF;
-    
-    /** Highest 52 bits, the key for the block. */
-    private static final long BLOCK_MASK = ~MID_MASK;
-    
-    /** Each block is 64 64-bit longs. */ 
-    private Map<Long, long[]> blocks = Maps.newHashMap();
 
-    private int index(long x) {
-        return (int)((x & MID_MASK) >> LOW_SHIFT);
-    }
+    private Map<Integer, RoaringBitmap> blocks = new HashMap<>();
 
-    private int bit(long x) {
-        return (int)(x & LOW_MASK);
+    public void add(long x) {
+        int high = highIndex(x);
+        int low = lowIndex(x);
+
+        if (!blocks.containsKey(high)) {
+            blocks.put(high, new RoaringBitmap());
+            LOG.debug("New block, {} blocks now", blocks.size());
+        }
+        blocks.get(high).add(low);
     }
-    
-    /** 
-     * The create parameter allows us to examine a block for a set bit 
-     * without creating an empty block as a side effect.
-     */
-    private long[] block(long x, boolean create) {
-        long key = x & BLOCK_MASK;
-        long[] block = blocks.get(key);
-        if (block == null && create) {
-            block = new long[64];
-            blocks.put(key, block);
-            LOG.debug("Number of blocks is now {}, total {} MiB", blocks.size(), blocks.size() * 64 * 8 / 1024.0 / 1024.0);
+
+    public boolean contains(long x) {
+        int high = highIndex(x);
+        RoaringBitmap block = blocks.get(high);
+        if (block != null) {
+            int low = lowIndex(x);
+            return block.contains(low);
+        } else {
+            return false; // block not found
         }
-        LOG.debug("   block {}", key);
-        return block;
     }
-    
-    public void add(long x) {
-        LOG.debug("set node {}: index {} bit {}", x, index(x), bit(x));
-        block(x, true)[index(x)] |= (1L << bit(x));
+
+    public int cardinality () {
+        // cardinality is the sum of the cardinality of all member bitmaps
+        return blocks.values().stream()
+                .mapToInt(RoaringBitmap::getCardinality)
+                .sum();
     }
 
-    public boolean contains(long x) {
-        LOG.debug("get node {}: index {} bit {}", x, index(x), bit(x));
-        long[] block = block(x, false);
-        if (block == null) return false; 
-        return (block[index(x)] & (1L << bit(x))) != 0;
+    private static int highIndex (long key) {
+        return (int) (key >> 32);
+    }
+
+    private static int lowIndex (long key) {
+        return (int) key;
```

Comment:
For all IDs with bit 31 set this is going to yield negative integers. https://docs.oracle.com/javase/specs/jls/se7/html/jls-5.html#jls-5.1.3 says "A narrowing conversion of a signed integer to an integral type T simply discards all but the n lowest order bits, where n is the number of bits used to represent type T. In addition to a possible loss of information about the magnitude of the numeric value, this may cause the sign of the resulting value to differ from the sign of the input value." Does the RoaringBitmap support storing negative integers?